store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedAccessHook.java

```diff
@@ -0,0 +1,126 @@
+ * <tr><td>  'A', 'B' </td><td>|</td><td> 'A'     </td><td>|</td><td> Y/N </td><td>|</td><td> Y   </td></tr>
+ * <tr><td>  'A'    </td><td>|</td><td> 'A', 'B'  </td><td>|</td><td> Y/N </td><td>|</td><td> Y   </td></tr>
+ * <tr><td>  'A'    </td><td>|</td><td> 'B'     </td><td>|</td><td> N   </td><td>|</td><td> N   </td></tr>
+ * <tr><td>  'A'    </td><td>|</td><td> 'B'     </td><td>|</td><td> Y   </td><td>|</td><td> Y   </td></tr>
+ * </table>
+ */
+public class FederatedAccessHook implements GraphHook {
+    public static final String USER_DOES_NOT_HAVE_CORRECT_AUTHS_TO_ACCESS_THIS_GRAPH_USER_S = "User does not have correct auths to access this graph. User: %s";
+    private Set<String> graphAuths;
+    private String addingUserId;
+
+    void setAddingUserId(final String creatorUserId) {
+        this.addingUserId = creatorUserId;
+    }
+
+    @Override
+    public void preExecute(final OperationChain<?> opChain, final User user) {
+        if (!isValidToExecute(user)) {
+            throw new FederatedAccessException(String.format(USER_DOES_NOT_HAVE_CORRECT_AUTHS_TO_ACCESS_THIS_GRAPH_USER_S, user.toString()));
+        }
+    }
+
+    public boolean isValidToExecute(final User user) {
+        final boolean isAddingUser = null != user.getUserId() && user.getUserId().equals(addingUserId);
+        final boolean authsIsEmpty = this.graphAuths != null && this.graphAuths.isEmpty();
+        final boolean userHasASharedAuth = this.graphAuths != null && !Collections.disjoint(user.getOpAuths(), this.graphAuths);
+
+        return authsIsEmpty
+                || isAddingUser
+                || userHasASharedAuth;
+    }
+
+    public FederatedAccessHook setGraphAuths(final Set<String> graphAuths) {
+        this.graphAuths = graphAuths;
+        return this;
+    }
+
+    @Override
+    public <T> T postExecute(final T result, final OperationChain<?> opChain, final User user) {
+        return result;
+    }
+
+    public static class Builder {
+        private FederatedAccessHook hook = new FederatedAccessHook();
+        private Builder self = this;
+
+        public Builder graphAuths(final String... opAuth) {
+            if (opAuth == null) {
+                hook.setGraphAuths(null);
+            } else {
+                graphAuths(Arrays.asList(opAuth));
+            }
+            return self;
+        }
+
+        public Builder graphAuths(final Collection<? extends String> opAuths) {
+            final HashSet<String> graphAuths = Sets.newHashSet(opAuths);
+            graphAuths.remove(null);
+            graphAuths.remove("");
+            if (hook.graphAuths == null) {
+                hook.graphAuths = Sets.newHashSet();
+            }
+            hook.setGraphAuths(graphAuths);
+            return self;
+        }
+
+        public FederatedAccessHook build() {
+            return hook;
+        }
+    }
+
+    public class FederatedAccessException extends RuntimeException {
```

Comment:
This should be in a separate file and it might be better as a checked exception.

Reply:
As discussed, would need to change signature of preExecute()